### PR TITLE
Refactors air alarm TLVs to be lazily initialized + general improvements.

### PR DESCRIPTION
**Pull request by vincentiusvin**
## About The Pull Request
Title really.

Lazy initialization mostly handled by set, reset, and kill value. As long as the return value is reassigned we should be good.
Also a bunch of documentation.
And split the big file into smaller chunks.
And type checking.
And automatic shutdown of pressure checks. ~~Actually I think it's probably better to set the ui minValue to 1 so accidental changes can get corrected quickly. Will do this later.~~ Changed my mind on this, its not a big deal in game as i thought. Might as well use the opportunity to tell folks that internal 0 = internal off.
And removed redundant 11 year old code (soulless PR)

PR-ing so the core logic can get reviewed since it's already done. A few extra stuffs regarding circuits and maps that I'll get to later. Smaller potatoes.

## Why It's Good For The Game
Better code (debatable)

## Changelog
:cl:
refactor: Refactored air alarm's threshold table list's logic. Yeah, that page that you never opened.
qol: A bunch more info on the air alarm UI.
qol: Made the pressure checks on vents shut down automatically on specific cases to avoid ambiguity.
code: Air alarm's UI code got changed a bit. No major change expected except from the aforementioned things. Report if bug.
del: Removed the air alarm circuit because it doesn't do anything useful.
/:cl:

### CODE REVIEW

File: tgui/packages/tgui/interfaces/AirAlarm.tsx [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+  hazard_max: number;
+};
+type AirAlarmDynamicData = {
+  locked: boolean;
+  siliconUser: boolean;
+  emagged: boolean;
+  danger_level: number;
+  atmos_alarm: boolean;
+  fire_alarm: boolean;
+  sensor_reading: { [criteria: string]: GasInformation };
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/AirAlarm.tsx:43]: You can use `Record<string, GasInformation>` here for less brackets and more semantics.

File: tgui/packages/tgui/interfaces/common/AtmosControls.tsx [hunk 2]
```diff
@@ -3,25 +3,46 @@ import { useBackend } from '../../backend';
 import { Button, LabeledList, NumberInput, Section } from '../../components';
 import { getGasLabel } from '../../constants';
 
-export const Vent = (props, context) => {
+export type VentData = {
+	id_tag: string;
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/common/AtmosControls.tsx:7]: Mixing tabs and spaces, please run a linter with --fix. Or just fix your vscode setup (Ctrl+Shift+P, Typescript Version, choose 4.3.5-sdk), after that saving the file automatically fixes it.

File: tgui/packages/tgui/interfaces/common/AtmosControls.tsx [hunk 3]
```diff
@@ -112,7 +142,7 @@ export const Vent = (props, context) => {
 };
 
 
-export const Scrubber = (props, context) => {
+export const Scrubber = (props:{scrubber:ScrubberData}, context) => {
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/common/AtmosControls.tsx:145]: Refactor that type into `type ScrubberProps`, this is how many ts devs do it, inlining is usually a horrible idea.

Comment (vincentiusvin) [reply tgui/packages/tgui/interfaces/common/AtmosControls.tsx]: Will keep that in mind. Let me know if b1a7442d5ea1e0cd73274e67069b217cdee8ca3f is bad.

**Comment by vincentiusvin** [comment]
Removed the air alarm circuit because it doesn't do what the PR advertises. All it does is set the TLV. Also passes the reference to the TLV datum when we are supposed to have numbers there.

Can bring it back sanely obviously, would prefer to do this in a different pr. (and maybe no gbp update 😉).

**Comment by vincentiusvin** [comment]
Bit more than i could chew. Putting this for another day.

**Comment by vincentiusvin** [comment]
Looks like the repathing is easier than i thought.

**Comment by vincentiusvin** [comment]
And redid mapped thermomachine logic too. Did you know that freezer thermomachines mins out the target temp roundstart but heater doesnt?

**Comment by LemonInTheDark** [comment]
Ready?

**Comment by vincentiusvin** [comment]
Aye

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Before I get into this, what did you change, and what did you just move?
(A commit of just moves would be best for this, but I realize it's hard to plan that stuff out ahead of time)

File: code/__DEFINES/machines.dm [hunk 4]
```diff
@@ -202,11 +202,19 @@
...
+#define AIR_ALARM_THRESHOLD_HAZARD_MAX (1 << 3)
+
+#define AIR_ALARM_THRESHOLD_SAFE 0
+#define AIR_ALARM_THRESHOLD_WARNING 1
+#define AIR_ALARM_THRESHOLD_HAZARDOUS 2
+
+/// Turn off the pump if x pressure has been reached on the tile.
+#define VENT_PUMP_EXT_BOUND (1 << 0)
+/// Turn off the pump if x pressure has been reached on the tile.
+#define VENT_PUMP_INT_BOUND (1 << 1)
```
[the remark this answers is not in the document]
Comment (vincentiusvin) [reply code/__DEFINES/machines.dm:220]: it's er, this thing
image, the internal and external bound. Shitty comment by me though ill fix it.

**Comment by vincentiusvin** [comment]
You probably can safely ignore the airalarm tool interaction, made no changes there except maybe var renaming.
The rest are modified quite significantly, except for the main airalarm file where a bunch are rewritten and a bunch also stay the same. You can check the airalarm changes in b50ab3c8fe770778f04ffd557f3ef423d4f60f87.

Worked on this before boon's firelock PR I'll also double check if his changes are still in later. You can put this off for now if you're also busy, cant get to reviews very fast for a week or two.

**Comment by vincentiusvin** [comment]
Actually now that i reread it will also need to make the string->path conversion for the tlvs to be similar for all kill, set, and reset. A bit all over the place right now. Will do this later too.

**Comment by vincentiusvin** [comment]
Local coder disappointed by own code. More at 11.

**Comment by timothymtorres** [comment]
This is great.  I had code dived into air alarms when I was adding GAGS to thermomachines and gas meters then decided fuck that.  The code was so old and clunky it was going to be a big PIA.  After this gets added I might give it another try!

**Comment by vincentiusvin** [comment]
Third time's the charm

**Comment by vincentiusvin** [comment]
Should be ready. Will reupdate paths once the PR in the linked issue is merged.

File: code/modules/atmospherics/machinery/airalarm/airalarm.dm [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+	idle_power_usage = BASE_MACHINE_IDLE_CONSUMPTION * 0.05
+	active_power_usage = BASE_MACHINE_ACTIVE_CONSUMPTION * 0.02
+	power_channel = AREA_USAGE_ENVIRON
+	req_access = list(ACCESS_ATMOSPHERICS)
+	max_integrity = 250
+	integrity_failure = 0.33
+	armor = list(MELEE = 0, BULLET = 0, LASER = 0, ENERGY = 100, BOMB = 0, BIO = 0, FIRE = 90, ACID = 30)
+	resistance_flags = FIRE_PROOF
+
+	var/danger_level = 0
```
Comment (Ghilker) [inline code/modules/atmospherics/machinery/airalarm/airalarm.dm:27]: doc these

File: code/modules/atmospherics/machinery/airalarm/airalarm.dm [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+	. = ..()
+	switch(build_stage)
+		if(AIRALARM_BUILD_NO_CIRCUIT)
+			. += span_notice("It is missing air alarm electronics.")
+		if(AIRALARM_BUILD_NO_WIRES)
+			. += span_notice("It is missing wiring.")
+		if(AIRALARM_BUILD_COMPLETE)
+			. += span_notice("Right-click to [locked ? "unlock" : "lock"] the interface.")
+
+/obj/machinery/airalarm/proc/reset(wire)
```
Comment (Ghilker) [inline code/modules/atmospherics/machinery/airalarm/airalarm.dm:92]: put docs on all these (args too)

File: code/modules/atmospherics/machinery/airalarm/airalarm.dm [hunk 7]
```diff
@@ -0,0 +1,586 @@
...
+			for(var/device_id in my_area.air_scrub_info)
+				send_signal(device_id, list(
+					"power" = 1,
+					"widenet" = 1,
+					"scrubbing" = 0
+				), signal_source)
+			for(var/device_id in my_area.air_vent_info)
+				send_signal(device_id, list(
+					"power" = 0
+				), signal_source)
```
Comment (Ghilker) [inline code/modules/atmospherics/machinery/airalarm/airalarm.dm:223]: these 1 list ones might be fine to be inline

File: code/modules/atmospherics/machinery/airalarm/airalarm.dm [hunk 8]
```diff
@@ -0,0 +1,586 @@
...
+/obj/machinery/airalarm/update_appearance(updates)
+	. = ..()
+
+	if(panel_open || (machine_stat & (NOPOWER|BROKEN)) || shorted)
+		set_light(0)
+		return
+
+	var/area/our_area = get_area(src)
+	var/color
+	switch(max(danger_level, !!our_area.active_alarms[ALARM_ATMOS]))
```
Comment (Ghilker) [inline code/modules/atmospherics/machinery/airalarm/airalarm.dm:266]: what's with the double not?

File: code/modules/atmospherics/machinery/airalarm/airalarm.dm [hunk 9]
```diff
@@ -0,0 +1,586 @@
...
+			if(AIRALARM_BUILD_COMPLETE)
+				icon_state = "alarmx"
+			if(AIRALARM_BUILD_NO_WIRES)
+				icon_state = "alarm_b2"
+			if(AIRALARM_BUILD_NO_CIRCUIT)
+				icon_state = "alarm_b1"
+		return ..()
+
+	icon_state = "alarmp"
+	return ..()
```
Comment (Ghilker) [inline code/modules/atmospherics/machinery/airalarm/airalarm.dm:288]: early return?

File: code/modules/atmospherics/machinery/airalarm/airalarm_tool_interactions.dm [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+	var/obj/item/wallframe/airalarm/alarm_frame = new(drop_location())
+	user.put_in_hands(alarm_frame)
+	qdel(src)
+	return TRUE
+
+/obj/machinery/airalarm/attackby(obj/item/W, mob/user, params)
+	update_last_used(user)
+	switch(build_stage)
+		if(AIRALARM_BUILD_COMPLETE)
+			if(W.GetID())// trying to unlock the interface with an ID card
```
Comment (Ghilker) [inline code/modules/atmospherics/machinery/airalarm/airalarm_tool_interactions.dm:51]: proper naming

File: code/modules/atmospherics/machinery/airalarm/airalarm_tool_interactions.dm [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+				return
+		if(AIRALARM_BUILD_NO_WIRES)
+			if(istype(W, /obj/item/stack/cable_coil))
+				var/obj/item/stack/cable_coil/cable = W
+				if(cable.get_amount() < 5)
+					to_chat(user, span_warning("You need five lengths of cable to wire the air alarm!"))
+					return
+				user.visible_message(span_notice("[user.name] wires the air alarm."), \
+									span_notice("You start wiring the air alarm..."))
+				if (do_after(user, 20, target = src))
```
Comment (Ghilker) [inline code/modules/atmospherics/machinery/airalarm/airalarm_tool_interactions.dm:65]: early returns on the istype and do_after, also why split the visible message?

File: code/modules/atmospherics/machinery/airalarm/airalarm_tool_interactions.dm [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+	if(machine_stat & (NOPOWER|BROKEN))
+		to_chat(user, span_warning("It does nothing!"))
+	else
+		if(src.allowed(usr) && !wires.is_cut(WIRE_IDSCAN))
+			locked = !locked
+			to_chat(user, span_notice("You [ locked ? "lock" : "unlock"] the air alarm interface."))
+			updateUsrDialog()
+		else
+			to_chat(user, span_danger("Access denied."))
+	return
```
Comment (Ghilker) [inline code/modules/atmospherics/machinery/airalarm/airalarm_tool_interactions.dm:133]: i think you could make this better with early returns

File: code/modules/atmospherics/machinery/airalarm/airalarm_variants.dm [hunk 13]
```diff
@@ -0,0 +1,47 @@
+/obj/machinery/airalarm/chilly_room
```
Comment (Ghilker) [inline code/modules/atmospherics/machinery/airalarm/airalarm_variants.dm:1]: cold room is way better

File: code/modules/atmospherics/machinery/components/unary_devices/thermomachine.dm [hunk 14]
```diff
@@ -36,6 +36,7 @@
 /obj/machinery/atmospherics/components/unary/thermomachine/Initialize(mapload)
 	. = ..()
 	RefreshParts()
+	target_temperature = clamp(target_temperature, min_temperature, max_temperature)
```
Comment (Ghilker) [inline code/modules/atmospherics/machinery/components/unary_devices/thermomachine.dm:39]: why is this necessary?

File: code/modules/atmospherics/machinery/components/unary_devices/thermomachine.dm [hunk 15]
```diff
@@ -291,29 +292,20 @@
...
-/obj/machinery/atmospherics/components/unary/thermomachine/freezer/on/coldroom/Initialize(mapload)
-	. = ..()
-	target_temperature = COLD_ROOM_TEMP
+	target_temperature = CHILLY_ROOM_TEMP
 
-/obj/machinery/atmospherics/components/unary/thermomachine/heater
+// Clamped on init, dont worry about these values.
+/obj/machinery/atmospherics/components/unary/thermomachine/on/coldest
+	name = "Cold room temperature control unit"
+	target_temperature = TCMB
```
Comment (Ghilker) [inline code/modules/atmospherics/machinery/components/unary_devices/thermomachine.dm:305]: ah now i see why, kinda weird you're doing it like this

**Comment by vincentiusvin** [comment]
Ignored some of the doc things, those aren't code I write and thus out of scope, just moved em around.

File: code/modules/atmospherics/machinery/airalarm/airalarm.dm (hunk 5 above)
Comment (vincentiusvin) [reply code/modules/atmospherics/machinery/airalarm/airalarm.dm:27]: ok

File: code/modules/atmospherics/machinery/airalarm/airalarm_variants.dm (hunk 13 above)
Comment (vincentiusvin) [reply code/modules/atmospherics/machinery/airalarm/airalarm_variants.dm:1]: 👍
